### PR TITLE
268/Data source developer docs

### PR DESCRIPTION
**Pull request by JenMadiedo**
Issue Number: 268
## Summary
Expanding on data source developer docs

## Changes
Adding docs about adding a new data source and how to
- Inherit from the data source interface
- Configure for dependency injection
- Register for sdk
- Add tests

Note, this PR will eventually go directly into the GSF repo where this file already exists in reduced form.

## Checklist

- [ ] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [X] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?
If yes, what are the expected API Changes? Please link to an API-Comparison workflow with the API Diff.

## Is this a breaking change?
If yes, what workflow does this break? 

## Anything else?
Other comments, collaborators, etc.
This PR Closes Issue #

### CODE REVIEW

**Comment by JenMadiedo** [comment]
Wasn't sure where/if I should include info about the logging infrastructure that currently exists, which can be leveraged in the data source or if that's too detailed. One option is to have and "Advanced" section where we include that info so it's not something we suggest as a first pass adding a data source but as a thing you can look into later. @juzuluag what do you think given you worked on that (I believe there were some docs written too)

**Comment by juzuluag** [comment]
> Wasn't sure where/if I should include info about the logging infrastructure that currently exists, which can be leveraged in the data source or if that's too detailed. One option is to have and "Advanced" section where we include that info so it's not something we suggest as a first pass adding a data source but as a thing you can look into later. @juzuluag what do you think given you worked on that (I believe there were some docs written too)

I think it won't hurt to mention that any approach using DI would work. That is the way we do logging, by injecting the logger through that mechanism. No need imo to have like an advance section for it.

File: docs/architecture/data-sources.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Data Sources
+
+Data sources allow developers easily integrate different data providers into the carbon aware SDK ([WattTime](https://www.wattime.org), [ElectricityMap](https://static.electricitymap.org), etc) to be made available to all higher level user-interfaces (WebAPI, CLI, etc), while avoiding the details of how to interact with any specific provider.
+
+## Data Sources' Responsibility
+Data sources act as the data ingestion layer for the SDK, handling the retrieval of data from a given data provider. They contain specific knowledge about the data provider they access, such as flags used in requests, fields that come back in responses, special use cases etc. They also handle any external calls that must be made to access the data provider. While helper clients can be built to handle the these calls, only the data source should have access to, and knowledge of, that client.
```
Comment (juzuluag) [inline docs/architecture/data-sources.md]: typo: `the these calls`

File: docs/architecture/data-sources.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+- For example, the WattTimeDataSource has a reference to a private WattTimeClient within it's implementation. The WattTimeClient handles the HTTP GET/POST calls to WattTime and the data source invokes the client once it has processed the request, and then processes the response before returning a final result.
+
+### Aggregator <-> Data Source Contract
+In order for the SDK to support different data sources, there is a defined contract between the Aggregator layer and the DataSource layer. This contract defines what a request from the Aggregator to the DataSource must contain, and what type of response the Aggregator expects from the DataSource. This means that each data source is responsible for: 
+- Pre-processing the Aggregator request must to create the expected request for the data provider.
+- Post-processing the data provider result to create the expected response type for the Aggregator
```
Comment (juzuluag) [inline docs/architecture/data-sources.md]: Maybe add that the `Aggregator` act as the "Business Logic" of the application.

File: docs/architecture/data-sources.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+### Post-Processing Caveat
+Post-processing should only ensure the types are what is expected and to fix any inconsistencies or issues that may be known to that specific data source. This post-processing **should not** do any extra data operations beyond those required to fulfill the Aggregator request ( i.e., averaging, min/max ops etc.). In other words, the data source should only manipulate data for the aim of returning _valid*_ data in the boundaries requested by the Aggregator
+
+\* What constitutes _valid_ data varies between data sources. It may be the case that some data sources don't handle time boundaries well so  extra processing may be required to ensure the data returned is what the aggregator expects assuming it was any data source and that those edge cases would be handled properly.
+
+## Creating a New Data Source
+
+Each new data source should be a new dotnet project under the `CarbonAware.DataSources` namespace. This project should have a reference to the `CarbonAware` project, and include the `Microsoft.Extensions.DependencyInjection` package. It should also be added to the solution. We have provided a command snippet below:
```
Comment (juzuluag) [inline docs/architecture/data-sources.md]: This is great!! 👍

**Review by juzuluag — APPROVED** [review]
Just a few comments. but overall looks great. Thank you so much for doing this. 👍

File: docs/architecture/data-sources.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+- For example, the WattTimeDataSource has a reference to a private WattTimeClient within it's implementation. The WattTimeClient handles the HTTP GET/POST calls to WattTime and the data source invokes the client once it has processed the request, and then processes the response before returning a final result.
+
+### Aggregator <-> Data Source Contract
+In order for the SDK to support different data sources, there is a defined contract between the Aggregator layer and the DataSource layer. This contract defines what a request from the Aggregator to the DataSource must contain, and what type of response the Aggregator expects from the DataSource. This means that each data source is responsible for: 
+- Pre-processing the Aggregator request must to create the expected request for the data provider.
```
Comment (pritipath) [inline docs/architecture/data-sources.md]:
```suggestion
- Pre-processing the Aggregator request to create the expected request for the data provider.
```

**Review by pritipath — APPROVED** [review]

File: docs/architecture/data-sources.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
+# Data Sources
+
+Data sources allow developers easily integrate different data providers into the carbon aware SDK ([WattTime](https://www.wattime.org), [ElectricityMap](https://static.electricitymap.org), etc) to be made available to all higher level user-interfaces (WebAPI, CLI, etc), while avoiding the details of how to interact with any specific provider.
```
Comment (bderusha) [inline docs/architecture/data-sources.md]: The ElectricityMap link is broken.  They recently rebranded and are now `ElectricityMaps` Consider linking to https://www.electricitymaps.com/

File: docs/architecture/data-sources.md (hunk 4 above)
Comment (bderusha) [inline docs/architecture/data-sources.md]:
```suggestion
- Pre-processing any arguments passed to it from the Aggregator to create the expected request for the data provider.
```

File: docs/architecture/data-sources.md (hunk 2 above)
Comment (bderusha) [inline docs/architecture/data-sources.md]:
```suggestion
- Post-processing the data provider result to create the expected return type for the Aggregator
```

File: docs/architecture/data-sources.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+- For example, the WattTimeDataSource has a reference to a private WattTimeClient within it's implementation. The WattTimeClient handles the HTTP GET/POST calls to WattTime and the data source invokes the client once it has processed the request, and then processes the response before returning a final result.
+
+### Aggregator <-> Data Source Contract
+In order for the SDK to support different data sources, there is a defined contract between the Aggregator layer and the DataSource layer. This contract defines what a request from the Aggregator to the DataSource must contain, and what type of response the Aggregator expects from the DataSource. This means that each data source is responsible for: 
+- Pre-processing the Aggregator request must to create the expected request for the data provider.
+- Post-processing the data provider result to create the expected response type for the Aggregator
+
+### Post-Processing Caveat
```
Comment (bderusha) [inline docs/architecture/data-sources.md]:
```suggestion
#### Post-Processing Caveat
```

Consider nesting this section within the above section since they seem related in that way more than as siblings

File: docs/architecture/data-sources.md (hunk 3 above)
Comment (bderusha) [inline docs/architecture/data-sources.md]:
```suggestion
Each new data source should be a new dotnet project under the `CarbonAware.DataSources` namespace and corresponding directory. This project should have a reference to the `CarbonAware` project, and include the `Microsoft.Extensions.DependencyInjection` package. It should also be added to the solution. We have provided a command snippet below:
```

File: docs/architecture/data-sources.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+Each new data source is expected to come with a robust unit test suite that ensures that the main flows and edge cases are properly handled. This also ensures that the SDK can switch seamlessly between data sources and the experiences up the stack remains consistent and helpful to the user.
+
+The unit tests should be added as a new project under the data source's test directory: `CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/test`. Be sure to include a reference to the data source's project and add it to the solution. We have provided a command snippet below:
+
+```sh
+cd src
+dotnet new nunit --name CarbonAware.DataSources.MyNewDataSource.Tests -o CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/test
+dotnet sln CarbonAwareSDK.sln add CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/test/CarbonAware.DataSources.MyNewDataSource.Tests.csproj
+cd CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource
+dotnet add tests/CarbonAware.DataSources.MyNewDataSource.Tests.csproj reference src/CarbonAware.DataSources.MyNewDataSource.csproj
```
Comment (bderusha) [inline docs/architecture/data-sources.md]:
```suggestion
dotnet add test/CarbonAware.DataSources.MyNewDataSource.Tests.csproj reference src/CarbonAware.DataSources.MyNewDataSource.csproj
```

File: docs/architecture/data-sources.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+{
+    public static void AddMyNewDataSource(this IServiceCollection services)
+    {
+        // ... register your data source with the IServiceCollection instance
+    }
+}
+```
+
+### Register the New Data Source
+Once the data source's ServiceCollectionExtensions is configured, it can be registered as an available data source for the SDK. This is done in the `CarbonAware.DataSources.Registration` project. In particular, there is an equivalent `ServiceCollectionExtensions` file in the `Configuration` directory of that project. The public function `AddDataSourceService` contains a switch statement that reads in the environment variables for the SDK and selects the requested data source. To register your new data source, you must:
```
Comment (bderusha) [inline docs/architecture/data-sources.md]: Consider linking to the appropriate file somewhere in this paragraph

File: docs/architecture/data-sources.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+```sh
+cd src
+dotnet new nunit --name CarbonAware.DataSources.MyNewDataSource.Tests -o CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/test
+dotnet sln CarbonAwareSDK.sln add CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/test/CarbonAware.DataSources.MyNewDataSource.Tests.csproj
+cd CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource
+dotnet add tests/CarbonAware.DataSources.MyNewDataSource.Tests.csproj reference src/CarbonAware.DataSources.MyNewDataSource.csproj
+```
+### Try it Out!
+You are now ready to try out your new data source! Add a new environment variable that sets the data source and make sure to read it into the DataSource.Registration project. If you added a new CarbonIntensity data source, you can set the `CarbonIntensityDataSource` env variable:
```
Comment (bderusha) [inline docs/architecture/data-sources.md]: How does one `make sure to read it into the DataSource.Registration project` ?  I would recommend either adding directions or removing this line.  Also we should be clear about using the term `configuration setting` rather than `env var`.  `env vars` are one way to set `configuration settings`, but there are others.  The setting is what controls the app, not the `env var`.

```suggestion
You are now ready to try out your new data source!  If you added a new `ICarbonIntensityDataSource`, you can configure it using the `CarbonIntensityDataSource` setting:
```

**Review by bderusha — CHANGES_REQUESTED** [review]
This is great work :)  Just a few suggestions for enhancement.

File: docs/architecture/data-sources.md (hunk 8 above)
Comment (JenMadiedo) [reply docs/architecture/data-sources.md]: What do you mean the appropriate file? Where the `AddDataSourceService` lives?

Comment (bderusha) [reply docs/architecture/data-sources.md]: Yes.  Rather than describing how to find it, with a link this paragraph could become:

> Once the data source's ServiceCollectionExtensions is configured, it can be registered as an available data source for the SDK by adding to the switch statement found in the `AddDataSourceService` function of [this file](/src/link/to/file).

File: docs/architecture/data-sources.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+### Post-Processing Caveat
+Post-processing should only ensure the types are what is expected and to fix any inconsistencies or issues that may be known to that specific data source. This post-processing **should not** do any extra data operations beyond those required to fulfill the Aggregator request ( i.e., averaging, min/max ops etc.). In other words, the data source should only manipulate data for the aim of returning _valid*_ data in the boundaries requested by the Aggregator
+
+\* What constitutes _valid_ data varies between data sources. It may be the case that some data sources don't handle time boundaries well so  extra processing may be required to ensure the data returned is what the aggregator expects assuming it was any data source and that those edge cases would be handled properly.
+
+## Creating a New Data Source
+
+Each new data source should be a new dotnet project under the `CarbonAware.DataSources` namespace. This project should have a reference to the `CarbonAware` project, and include the `Microsoft.Extensions.DependencyInjection` package. It should also be added to the solution. We have provided a command snippet below:
+
+```sh
```
Comment (gfmatthews) [inline docs/architecture/data-sources.md:23]: AWESOME providing a shell example here

**Review by gfmatthews — APPROVED** [review]
Ship it!

**Comment by JenMadiedo** [comment]
Closing now that have comments and will re-open with GSF history
